### DISCUSSION TITLE
[WSL2] vsock connection to X410 fails

### DISCUSSION THREAD

**Discussion by cshei** [microsoft/WSL#4723] [General]
**Please fill out the below information:**
* Your Windows build number:  Microsoft Windows [Version 10.0.19033.1]

* What you're doing and what's happening:

Follow steps at https://token2shell.com/howto/x410/quick-testing-vsock-hyper-v-support-in-x410/ to add registry keys and launch X410 with vsock support.

In either Gentoo or Ubuntu 18.04 on WSL2:
$ socat -b65536 UNIX-LISTEN:/tmp/.X11-unix/X5,fork,mode=777 SOCKET-CONNECT:40:0:x0000x70170000x02000000x00000000 &
$ export DISPLAY=:5.0
$ xterm

* What's wrong / what should be happening instead: connection fails and kernel log shows an error.  Running the same steps in an Ubuntu 19.04 Hyper-V VM works successfully.

$ xterm
2019/12/01 08:45:21 socat[334] E connect(5, AF=40 AF=40 0x0000701700000200000000000000, 16): Connection timed out
xterm: Xt error: Can't open display: :5.0

$ dmesg

[   80.407004] ------------[ cut here ]------------
[   80.407005] unknown msgtype=23
[   80.407021] WARNING: CPU: 2 PID: 0 at drivers/hv/vmbus_drv.c:909 vmbus_on_msg_dpc+0xff/0x1c0
[   80.407022] Modules linked in:
[   80.407024] CPU: 2 PID: 0 Comm: swapper/2 Not tainted 4.19.84-microsoft-standard #1
[   80.407025] RIP: 0010:vmbus_on_msg_dpc+0xff/0x1c0
[   80.407026] Code: 00 83 fe 15 0f 86 5d ff ff ff 80 3d 3c 24 d9 00 00 0f 85 72 ff ff ff 48 c7 c7 85 93 22 82 c6 05 28 24 d9 00 01 e8 7b 7b 8b ff <0f> 0b e9 58 ff ff ff 48 8b 3d eb c9 aa 00 ba 20 01 00 00 be 20 00
[   80.407027] RSP: 0018:ffff888321c83f28 EFLAGS: 00010282
[   80.407027] RAX: 0000000000000000 RBX: ffff888321eb2000 RCX: ffffffff82434ec8
[   80.407028] RDX: 0000000000000001 RSI: 0000000000000092 RDI: 0000000000000247
[   80.407028] RBP: 0000000000000000 R08: 0000000000000024 R09: 0000000000000001
[   80.407028] R10: 0000000000000000 R11: 0000000000000001 R12: 0000000000000001
[   80.407029] R13: ffff888321eb2210 R14: 0000000000000040 R15: 0000000000000006
[   80.407030] FS:  0000000000000000(0000) GS:ffff888321c80000(0000) knlGS:0000000000000000
[   80.407030] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   80.407030] CR2: 000055c11e3c7408 CR3: 000000000240a003 CR4: 00000000003606a0
[   80.407032] Call Trace:
[   80.407034]  <IRQ>
[   80.407038]  tasklet_action_common.isra.21+0x5a/0x100
[   80.407045]  __do_softirq+0xde/0x2de
[   80.407046]  irq_exit+0xba/0xc0
[   80.407047]  hyperv_vector_handler+0x57/0x70
[   80.407049]  hyperv_callback_vector+0xf/0x20
[   80.407049]  </IRQ>
[   80.407050] RIP: 0010:native_safe_halt+0xe/0x10
[   80.407051] Code: 1f 44 00 00 e9 c5 fe ff ff 0f 0b eb 94 0f 0b e9 04 ff ff ff 90 90 90 90 90 90 90 90 e9 07 00 00 00 0f 00 2d 14 9e 5c 00 fb f4 <c3> 90 e9 07 00 00 00 0f 00 2d 04 9e 5c 00 f4 c3 90 90 0f 1f 44 00
[   80.407051] RSP: 0018:ffffc90000093ec8 EFLAGS: 00000246 ORIG_RAX: ffffffffffffff0c
[   80.407052] RAX: ffffffff81a3ca30 RBX: 0000000000000002 RCX: ffffffff82436290
[   80.407053] RDX: 0000000000008782 RSI: ffffffff82433ef8 RDI: 0000000000000082
[   80.407053] RBP: 0000000000000002 R08: 0000000000000002 R09: 00000012b971d6c2
[   80.407054] R10: 0000001e1b9929f6 R11: 0000000000000000 R12: ffff888107e33800
[   80.407054] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[   80.407055]  ? __cpuidle_text_start+0x8/0x8
[   80.407057]  default_idle+0x1c/0x140
[   80.407058]  do_idle+0xe5/0x110
[   80.407060]  cpu_startup_entry+0x6f/0x80
[   80.407061]  start_secondary+0x195/0x1d0
[   80.407062]  secondary_startup_64+0xa4/0xb0
[   80.407064] ---[ end trace ecf9bcf6815fbd08 ]---

**Comment by Biswa96** [comment]
WSL2 itself uses vsock. If there is an error in vsock implementation we wouldn't be able to use WSL2. There may be some error in the x410 program or their method. Can you provide the error in `strace -ff` log?

**Comment by cshei** [comment]
Here's an strace -ff of the socat process in WSL... like I said, the same thing works in a full Hyper-V VM, so I'd suspect something on the WSL2/Microsoft kernel side.

[trace.5984.txt](https://github.com/microsoft/WSL/files/3909139/trace.5984.txt)
[trace.6008.txt](https://github.com/microsoft/WSL/files/3909140/trace.6008.txt)

**Comment by bketelsen** [comment]
subscribe, this is a really useful feature of X410

**Comment by 128keaton** [comment]
Also subscribed

**Comment by fquinner** [comment]
For a X GUI users on laptops this is a big deal since this effectively enables your X11 apps to survive a full hibernate.

**Comment by EtreRe** [comment]
This is also a big deal with VPN connection. When switching from a VPN connection to an other the X11 apps don't survive…

**Comment by fquinner** [comment]
dmesg reports this when attempted for me (but only on the first attempt after wsl2 boot).

```
[   20.084030] ------------[ cut here ]------------
[   20.084071] unknown msgtype=23
[   20.084088] WARNING: CPU: 2 PID: 0 at drivers/hv/vmbus_drv.c:909 vmbus_on_msg_dpc+0xff/0x1c0
[   20.084089] Modules linked in:
[   20.084094] CPU: 2 PID: 0 Comm: swapper/2 Not tainted 4.19.104-microsoft-standard #1
[   20.084098] RIP: 0010:vmbus_on_msg_dpc+0xff/0x1c0
[   20.084099] Code: 00 83 fe 15 0f 86 5d ff ff ff 80 3d b8 01 d1 00 00 0f 85 72 ff ff ff 48 c7 c7 4d 08 25 82 c6 05 a4 01 d1 00 01 e8 0b 68 89 ff <0f> 0b e9 58 ff ff ff 48 8b 3d fb 20 a4 00 ba 20 01 00 00 be 20 00
[   20.084101] RSP: 0018:ffff8883f7883f28 EFLAGS: 00010282
[   20.084102] RAX: 0000000000000000 RBX: ffff8883f7aff000 RCX: ffffffff82440908
[   20.084103] RDX: 0000000000000001 RSI: 0000000000000092 RDI: 0000000000000247
[   20.084104] RBP: 0000000000000000 R08: 0000000000000024 R09: 0000000000000001
[   20.084105] R10: 0000000000000000 R11: 0000000000000001 R12: 0000000000000001
[   20.084106] R13: ffff8883f7aff210 R14: 0000000000000040 R15: 0000000000000006
[   20.084107] FS:  0000000000000000(0000) GS:ffff8883f7880000(0000) knlGS:0000000000000000
[   20.084108] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   20.084109] CR2: 00007f72e4a1b858 CR3: 000000000240a006 CR4: 00000000003606a0
[   20.084112] Call Trace:
[   20.084115]  <IRQ>
[   20.084120]  tasklet_action_common.isra.21+0x5a/0x100
[   20.084124]  __do_softirq+0xde/0x2de
[   20.084127]  irq_exit+0xba/0xc0
[   20.084129]  hyperv_vector_handler+0x5f/0x80
[   20.084133]  hyperv_callback_vector+0xf/0x20
[   20.084135]  </IRQ>
[   20.084137] RIP: 0010:native_safe_halt+0xe/0x10
[   20.084139] Code: 1f 44 00 00 e9 c5 fe ff ff 0f 0b eb 94 0f 0b e9 04 ff ff ff 90 90 90 90 90 90 90 90 e9 07 00 00 00 0f 00 2d c4 74 54 00 fb f4 <c3> 90 e9 07 00 00 00 0f 00 2d b4 74 54 00 f4 c3 90 90 0f 1f 44 00
[   20.084140] RSP: 0018:ffffc90000093ec8 EFLAGS: 00000246 ORIG_RAX: ffffffffffffff0c
[   20.084141] RAX: ffffffff81ad1360 RBX: 0000000000000002 RCX: ffffffff82441cd0
[   20.084142] RDX: 0000000000005fde RSI: ffffffff8243f938 RDI: 0000000000000086
[   20.084143] RBP: 0000000000000002 R08: 0000000000000002 R09: 00fcd6c2a9ceee00
[   20.084144] R10: 0000000c1befe0a4 R11: 0000000000000000 R12: ffff888107e43800
[   20.084145] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[   20.084148]  ? __cpuidle_text_start+0x8/0x8
[   20.084151]  default_idle+0x1c/0x140
[   20.084155]  do_idle+0xe5/0x110
[   20.084158]  cpu_startup_entry+0x6f/0x80
[   20.084161]  start_secondary+0x195/0x1d0
[   20.084164]  secondary_startup_64+0xa4/0xb0
[   20.084167] ---[ end trace 73c4facebebdb3ca ]---
```
So at this point, socat is already started, DISPLAY is set for xterm, then this is triggered the first time xterm tries to connect over socat -> vsock.

Looking at the code in upstream, the enum does go up to 23:

https://github.com/torvalds/linux/blob/master/include/linux/hyperv.h#L405

But in the Microsoft WSL fork it only goes up to 21:

https://github.com/microsoft/WSL2-Linux-Kernel/blob/master/include/linux/hyperv.h#L408

So I wonder if this change is missing:

https://github.com/torvalds/linux/commit/ddc9d357b991838c2d975e8d7e4e9db26f37a7ff

**Comment by cshei** [comment]
I tried building a 5.4 kernel with the patch @fquinner mentioned, and while the error in dmesg is gone, I still get the error below when trying to open an X11 app, so it seems like vsock is really not supported in WSL2.  It does work for me in a real Hyper-V VM, so X410 is definitely set up correctly.

`2020/06/19 20:51:18 socat[149] E connect(5, AF=40 AF=40 0x0000701700000200000000000000, 16): Connection timed out`
`xterm: Xt error: Can't open display: :5.0`

**Comment by WSLUser** [comment]
I just got a suspicion about this. In the 4.19x series, vhost_vsock is disabled in the config due to a bug. However the bug is fixed in the 5.x series.  Enable `CONFIG_VHOST_VSOCK=y` on 5.4x or higher series and see if that does the trick. You can just add it on `.wslconfig` if you prefer to leave the default config when compiling the kernel. Alternatively you could use my [config](https://github.com/microsoft/WSL2-Linux-Kernel/blob/271456763a85394e318b345744ad8fd2692678c2/Microsoft/config-wsl) which has this option already enabled and requires 5.7.8 or newer kernel.

**Comment by cshei** [comment]
@WSLUser that appears to be for a VM host kernel to provide AF_VSOCK sockets for communicating with guests, so it seems to be unrelated.  In any case, I tried enabling it on a 5.4 kernel and it just causes WSL to not start... rebuilding the exact same kernel with only that one config option disabled resulting in a working kernel.

**Comment by WSLUser** [comment]
Strange. It works fine on the stable edition of the kernel. Are you also on Insiders? I'm just using 2004 with my config options and I have no difficulties whatsoever. In fact it feels better than normal. The crash is likely due to some other virtualization options not being enabled. The config option was originally enabled when WSL2 first appeared but was dropped due to issues that were later fixed. Those fixes never trickled down to the 4.19x LTS edition. See using my config as linked in my previous comment with the stable kernel. Be sure to run `make menu config` to get all the new options to appear. I just accepted defaults for the most part.

**Comment by TBBle** [comment]
The failures coming back are all "nothing listening on that port"; the "unknown msgtype" failure is just noise that can be ignored, since it turns into a timeout eventually, and the patch just silences this noise (which is a shame, since "port closed" would be a useful distinction for this case, versus "something there but it was too slow" timeouts).

The traces show it trying to connect to vsock port 6000 (0x00001770) (i.e. display 0), so you may want to check on the host side exactly what hyper-v address is being listened for, and that the registry entry for display 0 is present.

Based on [MS Docs](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/user-guide/make-integration-service#bind-to-a-hyper-v-socket) it should be a `SOCKADDR_HV` with `VmId` "90db8b89-0d35-4f79-8ce9-49ea0ac8b7cd" (`HV_GUID_CHILDREN`) (all-zeroes `HV_GUID_WILDCARD` would work too, I assume), and `ServiceId` "00001770-facb-11e6-bd58-64006a7986d3".

I'm not sure _how_ you check that, something that lists listening sockets might support `SOCKADDR_HV` or might be able to show you the raw sockaddr data. I haven't really looked around.

**Comment by fquinner** [comment]
The listen address and registry entries are correct since I can literally fire up a hyperv vm on same host and it works fine (I think others are doing the same). I did recompile latest upstream kernel and (after being amazed that it worked out of the box) it had no dmesg errors but still failed some weeks ago. I could try again with this configuration though I guess.

**Comment by TBBle** [comment]
What if you run your Hyper-V VM with the kernel used in WSL2? Assuming no other problems come up, that would rule-in or rule-out a kernel-related issue.

**Comment by fquinner** [comment]
@WSLUser I tried compiling kernel with your configuration and running it, but it did not work. Note that I see this which suggests that the kernel module is statically built in if i understand correctly (lsmod just returns an empty list on WSL since WSL config seems to favour static compilation and I remember last time i fought with this, trying to set up the hv modules as shared objects resulted on hanging on startup):

```
$ cat /lib/modules/5.7.12-microsoft-standard/modules.builtin | grep hv
kernel/drivers/scsi/hv_storvsc.ko
kernel/drivers/net/hyperv/hv_netvsc.ko
kernel/drivers/hv/hv_vmbus.ko
kernel/drivers/hv/hv_utils.ko
kernel/drivers/hv/hv_balloon.ko
kernel/net/vmw_vsock/hv_sock.ko
```

**Comment by fquinner** [comment]
I grabbed the kernel config from the working VM which was running Ubuntu 20.04 from `/boot/config-5.4.0-42-generic`, and built the 5.4.0 kernel on WSL2 with this build configuration file to try and run a known working kernel for hyper-v on wsl2 but it failed to boot. Presumably some WSL2 magic is required that is missing from that kernel configuration? Or is there some critical-to-bootstrapping reason that the WSL2 kernel builds all modules statically? I would have tried the reverse with a VM, but I don't really have the time to fight with however grub works with that distro.

**Comment by WSLUser** [comment]
You still need to run the make modules command, which will append the "Microsoft standard" to the kernel. It shouldn't be needed but something in WSL2 could be checking for it. Module support is limited. You could install much of it as modules but you'd need to use modprobe everytime. Everything is just built-in as that's how MS currently does it. When they have better module support, hopefully we can switch a bunch of it to =m. I have it working but it was before the last 2 commits I did. But they just add more support for things dealing with 9P (some bugs in Insiders I guess) and perf testing. I plan on updating to latest upstream soon so I'll have an updated config with any new options. Also try doing it as a .config and pure make instead of kconfig. I used Nunix's ZFS guide as it provides steps to compile kernel easily.

**Comment by fquinner** [comment]
Thanks @WSLUser I did run the `make modules_install` command ahead of trying this. I was loosely following Hayden's guide at https://boxofcables.dev/accelerated-kvm-guests-on-wsl-2/ and using the `.config` method but really it doesn't seem like there's much to it. When you say you have it working, do you mean with X410 and your `DISPLAY` is definitely set to the socat source and not the old TCP/IP address? Or that you have another vsock connection running and working? I did wonder if there was a conflict with the vsock address but X410 is closed source so I can't modify it and it seems a non trivial thing to test with like a ping pong vsock server or something like that.

**Comment by WSLUser** [comment]
No I just mean the kernel works fine. Not that x410 works. I use vcxsrv. Not sure if they support hyper-v sockets but if they do, they're OSS at least

**Comment by fquinner** [comment]
Ah right yeah the kernel does work fine when using latest kernel but vsock doesn't seem to play nicely with it. Vsock I think is only available on X410 and is super nice if it works because it means your x sessions can survive hibernate, suspend, network change etc. It's weird though because the kernel does look like it contains it OK but it doesn't seem to be able to work with X410 / wsl2.

**Comment by TBBle** [comment]
According to https://github.com/microsoft/WSL2-Linux-Kernel/issues/7, at some point, WSL2's own startup process is using VSock to connect to systems on the host. So it seems unlikely that vsock itself is broken in the kernel build.

We can see that with `ss`.

```
root@KEITARO:~# ss -f vsock -p -la
Netid          State           Recv-Q           Send-Q                     Local Address:Port                           Peer Address:Port                Process
v_str          ESTAB           0                0                                      *:1325441449                                2:50000
v_str          ESTAB           0                0                                      *:1325441450                                2:50000
v_str          ESTAB           0                0                                      *:1325441451                                2:50001
v_str          ESTAB           0                0                                      *:1325441452                                2:50000
v_str          ESTAB           0                0                                      *:1325441453                                2:50000
v_str          ESTAB           0                0                                      *:1325441454                                2:50000                users:(("init",pid=1,fd=100))
v_str          LISTEN          0                0                                      *:1325441455                                *:*
v_str          ESTAB           0                0                                      *:1325441456                                2:50002
v_str          ESTAB           0                0                                      *:1325441457                                2:50002
v_str          LISTEN          0                0                                      *:1325441458                                *:*                    users:(("init",pid=9,fd=9),("init",pid=8,fd=9),("init",pid=1,fd=9))
v_str          ESTAB           0                0                                      *:1325441458                                2:3161828484           users:(("init",pid=9,fd=4),("init",pid=8,fd=4),("init",pid=1,fd=4))
v_str          ESTAB           0                0                                      *:1325441459                                2:3161828486           users:(("init",pid=9,fd=12),("init",pid=8,fd=12))
v_str          ESTAB           0                0                                      *:1325441460                                2:3161828490           users:(("init",pid=9,fd=16))
v_str          ESTAB           0                0                                      *:1325441460                                2:3161828489           users:(("init",pid=9,fd=15))
v_str          ESTAB           0                0                                      *:1325441460                                2:3161828488           users:(("init",pid=9,fd=14))
v_str          ESTAB           0                0                                      *:1325441460                                2:3161828487           users:(("init",pid=9,fd=13))
```

I wonder if the WSL2 uVM doesn't show up in a group such as `HV_GUID_CHILDREN`, if that's what x410 is listening on... Perhaps find something like socat for Windows, stop x410, and try listening on the relevant port (6000) with `HV_GUID_CHILDREN`, and start X and see if you get a connection. If that doesn't work, try `HV_GUID_WILDCARD`. If that works, raise a bug report with x410 that they need to listen on `HV_GUID_WILDCARD` or WSL2 won't work.

That's speculation though. I haven't found anything like `ss` for Windows that'll show listening `AF_HYPERV` sockets.

**Comment by ameeno** [comment]
Hello, I'm very interested in having X410 + WSL2 + Vsock display working.

Does anyone know how to achieve this?

I currently have Sound working fine as I have Pulse Server installed a service on windows side, and x410 installed as a service windows side. But sleep/hibernate breaks the display x410 connection.


If i was able to get it working under Sock, would the pulse server continue to work, or would i need to have an XRDP pulse sink?

**Comment by agowa** [comment]
When you use vsock and x410 there is no need for xrdp at all.

**Comment by TBBle** [comment]
Just to nail this down, I confirmed that _incoming_ VSock connections from Windows into the WSL2 session work fine with stock WSL2 Ubuntu and kernel `4.19.104-microsoft-standard`. I used the code from a [Stack Overflow question](https://stackoverflow.com/q/60696166/166389), fixed per my answer there, and was able to connect using that code to the microVM with the GUID gotten from `hcsdiag list`, to a copy of `nc-vsock` listening in my WSL2 Ubuntu session.

For tracing HyperV sockets, you can use [Winsock Tracing](https://docs.microsoft.com/en-us/windows/win32/winsock/control-of-winsock-tracing). "Family 34" in the resulting logs is `AF_HYPERV`, and the Address will be `0x22000000<VM GUID>`, which should be either `HV_GUID_CHILDREN` (albeit byte-swapped) or `HV_GUID_WILDCARD` (all zeroes) when X410 runs.

So I turned the code from Stack Overflow into a listener, and I think I have identified the issue. (Full Host-side code and test instructions inside the *Details* fold)

<details>

Fill in your MicroVM UUID and compile with `CL`, and run it, then connect from the WSL side with `nc-vsock 2 6000`, assuming you've built https://github.com/stefanha/nc-vsock. You can also use `socat stdin SOCKET-CONNECT:40:0:x0000x70170000x02000000x00000000`.

If you get 'connected timed out' on the WSL side, it didn't work. If the test exits immediately, and so does the listener on the Windows side, then it worked.

```cpp
#include <iostream>

#include <winsock2.h>
#include <ws2tcpip.h>
#include <hvsocket.h>

#include <combaseapi.h>

#pragma comment(lib, "Ws2_32.lib")

int main()
{
    // UUID FROM hcsdiag list
    struct __declspec(uuid("4D5065CF-9632-484E-A27D-C2F1B09CDBD3")) ServerVsockTemplate {};

    //----------------------
    // Initialize Winsock
    WSADATA wsaData;
    int iResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
    if (iResult != NO_ERROR) {
        wprintf(L"WSAStartup function failed with error: %d\n", iResult);
        return 1;
    }
    //----------------------
    // Create a SOCKET for connecting to server
    SOCKET ListenSocket;
    ListenSocket = socket(AF_HYPERV, SOCK_STREAM, HV_PROTOCOL_RAW);
    if (ListenSocket == INVALID_SOCKET) {
        wprintf(L"socket function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }
    //----------------------
    // The sockaddr_in structure specifies the address family,
    // IP address, and port of the server to be listened on.
    SOCKADDR_HV clientService;
    clientService.Family = AF_HYPERV;
    clientService.Reserved = 0;
    clientService.VmId = __uuidof(ServerVsockTemplate);
    clientService.ServiceId = HV_GUID_VSOCK_TEMPLATE;
    clientService.ServiceId.Data1 = 6000;

    //----------------------
    // Bind the socket
    iResult = bind(ListenSocket, (SOCKADDR*)&clientService, sizeof(clientService));
    if (iResult == SOCKET_ERROR) {
        wprintf(L"bind function failed with error: %ld\n", WSAGetLastError());
        iResult = closesocket(ListenSocket);
        if (iResult == SOCKET_ERROR)
            wprintf(L"closesocket function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }

    iResult = listen(ListenSocket, SOMAXCONN);
    if (iResult == SOCKET_ERROR) {
        wprintf(L"listen function failed with error: %ld\n", WSAGetLastError());
        iResult = closesocket(ListenSocket);
        if (iResult == SOCKET_ERROR)
            wprintf(L"closesocket function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }
    wprintf(L"Listening for clients.\n");

    SOCKET ConnectedSocket = accept(ListenSocket, nullptr, 0);
    if (ConnectedSocket == INVALID_SOCKET) {
        wprintf(L"accept function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        iResult = closesocket(ListenSocket);
        if (iResult == SOCKET_ERROR)
            wprintf(L"closesocket function failed with error: %ld\n", WSAGetLastError());
        return 1;
    }

    iResult = closesocket(ConnectedSocket);
    if (iResult == SOCKET_ERROR) {
        wprintf(L"closesocket(ConnectedSocket) function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }

    iResult = closesocket(ListenSocket);
    if (iResult == SOCKET_ERROR) {
        wprintf(L"closesocket(ListenSocket) function failed with error: %ld\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }

    WSACleanup();
    return 0;
}

```
</details>

While a normal Hyper-V VM can be listened-to using `HV_GUID_WILDCARD` or `HV_GUID_CHILDREN`, I was only able to get a VSock listener on the Host to be reachable from inside the WSL2 VM if I listened on the exact GUID of the WSL2 MicroVM.

i.e., this worked:
```
    SOCKADDR_HV clientService;
    clientService.Family = AF_HYPERV;
    clientService.Reserved = 0;

    // GUID from `hcsdiag list` (as Administrator) while the WSL2 VM was running
    struct __declspec(uuid("4D5065CF-9632-484E-A27D-C2F1B09CDBD3")) ServerVsockTemplate {};
    clientService.VmId = __uuidof(ServerVsockTemplate);

    clientService.ServiceId = HV_GUID_VSOCK_TEMPLATE;
    clientService.ServiceId.Data1 = 6000;
```

but this didn't work, which is almost certainly what X410 does internally. This version would work with any normal Hyper-V VM.

```
    SOCKADDR_HV clientService;
    clientService.Family = AF_HYPERV;
    clientService.Reserved = 0;

    clientService.VmId = HV_GUID_CHILDREN; // Or HV_GUID_WILDCARD

    clientService.ServiceId = HV_GUID_VSOCK_TEMPLATE;
    clientService.ServiceId.Data1 = 6000;
```

So something about the way the MicroVM is set up, it _doesn't_ see generically-listening Guest Communication Services. This is _probably_ a bug MS will need to deal with, as I'm unsure if this is intrinsic to the GCS system, or something about WSL that can be fixed. It might also be possible that there's a different known-UUID that can be listened-on, or a socket option that can be set, to make the HCS MicroVM able to connect to a socket that isn't bound to it specifically.

In the meantime, either X410 can be modified to be able to be told what HCS MicroVM to listen for, or (with Admin rights) it can search for one itself. Or perhaps a little utility could be written that listens for the MicroVM, and forwards to X410, rather than using X410's own VSock support.

I also checked, and I didn't need to add the registry entries from the [X410 docs](https://x410.dev/cookbook/hyperv/using-x410-with-hyper-v-linux-virtual-machines-via-vsock/). I suspect they are only needed when listening for `HV_GUID_CHILDREN` or `HV_GUID_WILDCARD`. Or perhaps for VSock they're implicit.

**Comment by fquinner** [comment]
@TBBle Ha you beat me to it I was going to try exactly that when I got back from holidays - nice work.

**Comment by TBBle** [comment]
One thing worth noting is that the WSL2 `/sbin/init` is connecting to a VSock on the host on port 50000, and I can also see connections to 50001 and 50002. However, those might be talking to services started by wsl or wslhost, which could be listening explicitly to the HCS MicroVM UUID, as I did above. That would make sense, as they know the HCS MicroVM UUID at start-up time.

Edit: Yup, confirmed with Winsock Tracing. From a shutdown state, a HCS MicroVM with UUID "2F34B9CA-408C-428D-AC02-9666646AE568" was created, and you can see a socket bind and listen:
```
		<Message>socket: 0: Process 0xFFFFA78DD9ABB080 (0x550C), Endpoint 0xFFFFA78DE10C5320, Family 34, Type SOCK_STREAM , Protocol 1, Seq 1006, Status STATUS_SUCCESS </Message>
		<Message>socket: 1: Process 0xFFFFA78DD9ABB080 (0x550C), Endpoint 0xFFFFA78DE10C5320, Family 0, Type 0, Protocol 0, Seq 1013, Status STATUS_SUCCESS </Message>
		<Message>bind: 0: Process 0xFFFFA78DD9ABB080, Endpoint 0xFFFFA78DE10C5320, Address 0x22000000CAB9342F8C408D42AC029666, Seq 7010, Status STATUS_SUCCESS </Message>
		<Message>bind: 1: Process 0xFFFFA78DD9ABB080, Endpoint 0xFFFFA78DE10C5320, Address 0x22000000898BDB90350D794F8CE949EA, Seq 7022, Status STATUS_SUCCESS </Message>
		<Message>Listen: 0: Process 0xFFFFA78DD9ABB080, Endpoint 0xFFFFA78DE10C5320, Backlog 200, Seq 13006, Status STATUS_SUCCESS </Message>
		<Message>Listen: 1: Process 0xFFFFA78DD9ABB080, Endpoint 0xFFFFA78DE10C5320, Backlog 200, Seq 13012, Status STATUS_SUCCESS </Message>
		<Message>AcceptEx: 0: Process 0xFFFFA78DD9ABB080, Endpoint 0xFFFFA78DE10C5320, Seq 6024, Status STATUS_SUCCESS </Message>
...
```

You can see the bind address 0x22000000CAB9342F8C408D42AC029666: "Family 34, Reserved 00, VmID 0xCAB9342F8C408D42AC029666 (== UUID '2F34B9CA-408C-428D-AC02-9666...' cut off due to trace output)" I'm not sure what that return address (0x898BDB90350D794F8CE949EA...) is, but I suspect that it's the UUID of my machine, and I'd see it if I were to query the listening socket with `getsockname`.

So if there's a magic "Listen to all VMs _and_ all HCS MicroVMs" GUID, the existing WSL2 code isn't using it.

**Comment by ameeno** [comment]
Great work guy's. SO as x410 is closed source, we need some sort of way in to modify x410 either via conf files?

Perhaps X410 dev can use your instruction to see how it can detect the GUID of the MicroVM on the host side and listen for it?

**Comment by fquinner** [comment]
I pinged the x410 support desk and referenced this ticket will see what they come back with...

**Comment by ameeno** [comment]
The WSL GUID changes every time the MicroVM is shutdown and restarted.

**Comment by agowa** [comment]
It's a bug. HV_GUID_WILDCARD and HV_GUID_CHILDREN are not called how they are without a reason.
Even if the X410 devs would implement a workaround we'd have similar issues if other developers want to use the same api.

And thanks @TBBle for figuring this out. I hope MS will be able to fix this soon for insiders.

**Comment by TBBle** [comment]
Yeah, sorry, I forgot to be explicit, that the GUIDs I used were just what happened to come up on my machine while testing: there's no fixed WSL2 MicroVM GUID. You can see the different GUIDs between my two comments. ^_^

I suspect the root cause here is that the HCS MicroVMs are also used for Windows Containers, and I am not sure if containers would be expected to see the same Guest Communication Services as VMs see, without something in the OCI setup specifying a shared network namespace, as it would be a silent isolation break. Hopefully this is something the WSL2 MicroVM can be configured to allow, without breaking the isolation of Windows Containers MicroVMs.

**Comment by ameeno** [comment]
That, or we use X over tcp for now and wait for MS to release the Gui Apps via wsl which is on the roadmap.

Because i think your right, Patching WSL2 to work on the standard Hyper-V stack would break isolation for windows containers.

**Comment by TBBle** [comment]
I just learned today (from [here](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md#implementation-details)) that WSL2-supporting Windows should also support Unix Domain Sockets (`AF_UNIX`) where a Windows Container can listen on a socket, and the host can connect to it.

> Support for Unix Domain Sockets has been introduced in Windows Server 2019 and works across containers as well as host and container as long as the processes running in containers are listening on the socket. If a process from within a container wishes to connect to a domain socket that a process on the host OS is listening on, Windows returns a permission error. This scenario however does not arise in the context of interactions between Kubelet, CSI Node Driver Registrar and CSI Node Plugin as these involve a process in a container listening on a domain socket (CSI Node Driver Registrar or CSI Node Plugin) that a process on the host (Kubelet) connects to.

If this also works for WSL2 micro VMs (which may be true for the same reason that that `AF_VSOCK` _doesn't_ work), then this would be a good solution to X11-style flows where we want the host to connect to the guest.

It's also possible that it won't work without doing some specific setup for the WSL2 microVM which we don't have control over anyway.

I haven't followed up to confirm this discovery with any other documentation, but if that turns out to be a reasonable workaround, it would simplify a bunch of stuff.

And poking around, https://github.com/microsoft/WSL/issues/4240#issuecomment-530495107 suggests this either doesn't, or didn't work for WSL2.

Edit: I did find documentation about how [Windows/WSL Interop with AF_UNIX](https://devblogs.microsoft.com/commandline/windowswsl-interop-with-af_unix/) used to work, at least.

**Comment by cshei** [comment]
Someone has come up with an implementation that listens on the WSL2 micro VM's GUID that seems to work quite well: https://github.com/nbdd0121/x11-over-vsock

Given that official GUI support is coming in the near future in WSL, this is probably a good enough workaround to avoid the sleep/resume disconnects until then.

**Reply by TBBle** [reply]
That's a neat solution. Note that the Linux-side of that is pretty much the same thing done with `socat` in the X410 instructions, so it's really only the Windows side that's important, as it implements the PoC windows-side code from https://github.com/microsoft/WSL/discussions/4723#discussioncomment-63061 and also provides a neat daemon mode.

The fact that daemon mode needs to be run with administrator privilege might be one reason X410 haven't implemented it.
